openmls/src/binary_tree/array_representation/tree.rs

```diff
@@ -0,0 +1,176 @@
+    type Error = ABinaryTreeError;
+
+    fn try_from(nodes: Vec<T>) -> Result<Self, Self::Error> {
+        Self::new(nodes)
+    }
+}
+
+impl<T: Clone + Debug> ABinaryTree<T> {
+    /// Create a tree from the given vector of nodes. The vector of nodes can't
+    /// be empty and has to yield a full, left-balanced binary tree. The nodes
+    /// in the tree are ordered in the array-representation. This function
+    /// throws a [`ABinaryTreeError::InvalidNumberOfNodes`] error if the number of nodes does not
+    /// allow the creation of a full, left-balanced binary tree and an
+    /// [`ABinaryTreeError::OutOfRange`] error if the number of given nodes exceeds the range of
+    /// [`NodeIndex`].
+    pub(crate) fn new(nodes: Vec<T>) -> Result<Self, ABinaryTreeError> {
+        if nodes.len()
+            > usize::try_from(NodeIndex::MAX).map_err(|_| ABinaryTreeError::LibraryError)?
+        {
+            return Err(ABinaryTreeError::OutOfRange);
+        }
+        if nodes.len() % 2 != 1 {
+            return Err(ABinaryTreeError::InvalidNumberOfNodes);
+        }
+        Ok(ABinaryTree { nodes })
+    }
+
+    /// Obtain a reference to the data contained in the node at index
+    /// `node_index`, where the indexing corresponds to the array representation
+    /// of the underlying binary tree. Returns [`ABinaryTreeError::OutOfBounds`]
+    /// if the index is larger than the size of the tree.
+    pub(in crate::binary_tree) fn node_by_index(
+        &self,
+        node_index: NodeIndex,
+    ) -> Result<&T, ABinaryTreeError> {
+        self.nodes
+            .get(usize::try_from(node_index).map_err(|_| ABinaryTreeError::LibraryError)?)
+            .ok_or(ABinaryTreeError::OutOfBounds)
+    }
+
+    /// Return the number of nodes in the tree.
+    pub(in crate::binary_tree) fn size(&self) -> Result<NodeIndex, ABinaryTreeError> {
+        let tree_size =
+            u32::try_from(self.nodes.len()).map_err(|_| ABinaryTreeError::LibraryError)?;
+        Ok(tree_size)
+    }
+
+    /// Return the number of leaves in the tree.
+    pub(crate) fn leaf_count(&self) -> Result<TreeSize, ABinaryTreeError> {
+        // This works, because the tree always has at least one leaf.
+        Ok(((self.size()? - 1) / 2) + 1)
+    }
+
+    /// Return a vector of leaves sorted according to their position in the tree
+    /// from left to right. This function should not fail and only returns a
+    /// [`Result`], because it might throw a
+    /// [`LibraryError`](ABinaryTreeError::LibraryError).
+    pub(crate) fn leaves(&self) -> Result<Vec<&T>, ABinaryTreeError> {
+        let mut leaf_references = Vec::new();
+        for leaf_index in 0..self.leaf_count()? {
+            let node_index = usize::try_from(to_node_index(leaf_index))
+                // The tree size and thus the leaf count should fit into usize on 32
+                // bit architectures.
+                .map_err(|_| ABinaryTreeError::LibraryError)?;
```

Comment:
Because you where complaining about having to use for loops.
You can always `collect::<Result<Vec<_>, _>>` and them handle the error later.